neo/Consensus/ConsensusService.cs

```diff
@@ -289,17 +348,141 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
+            {
+                Contract contract = Contract.CreateMultiSigContract(context.M, context.Validators);
+                Block block = context.MakeHeader();
+                if (block == null) return;
+                context.State |= ConsensusState.BlockSent;
+
+                ContractParametersContext sc = new ContractParametersContext(block);
+                for (int i = 0, j = 0; i < context.Validators.Length && j < context.M; i++)
+                    if (context.FinalSignatures[i] != null)
+                    {
+                        sc.AddSignature(contract, context.Validators[i], context.FinalSignatures[i]);
+                        j++;
+                    }
+                sc.Verifiable.Witnesses = sc.GetWitnesses();
+                block.Transactions = context.TransactionHashes.Select(p => context.Transactions[p]).ToArray();
+                Log($"relay block: height={context.BlockIndex} hash={block.Hash}");
+
+                system.LocalNode.Tell(new LocalNode.Relay { Inventory = block });
+            }
+        }
+
+        private void OnRenegeration(ConsensusPayload payload, Renegeration message)
+        {
+            Log($"{nameof(OnRenegeration)}: height={payload.BlockIndex} hash={context.MakeHeader().Hash.ToString()} view={message.ViewNumber} numberOfPartialSignatures={message.SignedPayloads.Count(p => p != null)} index={payload.ValidatorIndex}");
+
+            //
+            /// <summary>
+            /// Time for checking if speaker really signed this payload
+            /// </summary>
+
+            /*
+            // TODO - Remove signature from payload and check Primary
+            public byte[] PrimaryPrepReqSignature = message.SignedPayloads[context.PrimaryIndex];
+            message.PrepReqSignature = new byte[64];
+            payload.Data = message.ToArray();
+            if (!Crypto.Default.VerifySignature(payload.GetHashData(), context.SignedPayloads[payload.ValidatorIndex], context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+            {
+                context.SignedPayloads[payload.ValidatorIndex] = null;
+                return;
+            }
+            /// <summary>
+            /// These next 2 lines could be removed, because payload is not anymore used
+            /// it was already saved before changed in the context.PreparePayload... However, let keep things clean for now
+            /// </summary>
+            message.PrepReqSignature = context.SignedPayloads[payload.ValidatorIndex];
+            payload.Data = message.ToArray(); 
+
+            */
+
+
+            /// <summary>
+            /// Time for checking all Backups
+            /// </summary>
+            uint nValidSignatures = 0;
+            for (int i = 0; i < message.SignedPayloads.Length; i++)
+                if (message.SignedPayloads[i] != null && i != message.PrepareRequestPayload.ValidatorIndex)
```

Comment:
May I ask why the primary is excluded here ? 

I think primary's SignedPayload is also count as 1 in M signatures.

Reply:
Hi, @wanglongfei88,
The primary is not excluded. There is a TODO in code for inserting the piece of code for its calculus, I will do it today.

The point is that the Primary also signed the `PrepareRequest`, right? However, the `PrepareRequest` Payload that he signed did not have a signature. Variable `public byte[] PrepReqSignature;` was empty.
In this sense, we need to check the Primary signature in a different manner, I will try to update it right now!

Reply:
done. :dagger: 

Thanks for the valuable and precise review and suggestions.

Reply:
In addition, the function for checking his signature is not isolated:

```cs
        private bool CheckPrimaryPayloadSignature(ConsensusPayload payload, PrepareRequest message)
        {
            /// <summary>
            /// Partial signatures of, at least, M nodes
            /// The Speaker Signed the Payload without any signature (this was the trick/magic part)
            /// But the payload was modified with the signature after that. 
            /// Thus, we need to remove the signature from the Payload to correctly verify Speaker identity agreements with this block
            /// </summary>
            message.PrepReqSignature = new byte[64];
            payload.Data = message.ToArray();
            if (!Crypto.Default.VerifySignature(payload.GetHashData(), context.SignedPayloads[payload.ValidatorIndex], context.Validators[payload.ValidatorIndex].EncodePoint(false)))
                return true;

            /// <summary>
            /// These next 2 lines could be removed, because payload is not anymore used
            /// it was already saved before changed in the context.PreparePayload... However, let keep things clean for now
            /// </summary>
            message.PrepReqSignature = context.SignedPayloads[payload.ValidatorIndex];
            payload.Data = message.ToArray();
            return false;
        }
```

Reply:
Totally agree. Primary signs PreparePayload without signature. But other CNs sign it with primary's signature. So verification will be different and it is a good idea to separate it. 
Good job, Cheers.

Reply:
I am glad you like and got the message.